src/backends/github/__tests__/API.spec.js

```diff
@@ -1,39 +1,274 @@
+import fetchMock from 'fetch-mock';
+import { curry, escapeRegExp, isMatch, merge } from 'lodash';
+import { Map } from 'immutable';
+
 import AssetProxy from "../../../valueObjects/AssetProxy";
-import API from "../API";
+import API from '../API';
+
+const compose = (...fns) => val => fns.reduceRight((newVal, fn) => fn(newVal), val);
+const pipe = (...fns) => compose(...fns.reverse());
+
+const regExpOrString = rOrS => (rOrS instanceof RegExp ? rOrS.toString() : escapeRegExp(rOrS));
+
+const mockForAllParams = url => `${url}(\\?.*)?`;
+const prependRoot = urlRoot => url => `${urlRoot}${url}`;
+const matchWholeURL = str => `^${str}$`;
+const strToRegex = str => new RegExp(str);
+const matchURL = curry((urlRoot, forAllParams, url) => pipe(
+  regExpOrString,
+  ...(forAllParams ? [mockForAllParams] : []),
+  pipe(regExpOrString, prependRoot)(urlRoot),
+  matchWholeURL,
+  strToRegex,
+)(url));
+
+// `mock` gives us a few advantages over using the standard
+// `fetchMock.mock`:
+//  - Routes can have a root specified that is prepended to the path 
+//  - By default, routes swallow URL parameters (the GitHub API code
+//    uses a `ts` parameter on _every_ request)
+const mockRequest = curry((urlRoot, url, response, options={}) => {
+  const mergedOptions = merge({}, {
+    forAllParams: true,
+    fetchMockOptions: {},
+  }, options);
+  return fetchMock.mock(
+    matchURL(urlRoot, mergedOptions.forAllParams, url),
+    response,
+    options.fetchMockOptions,
+  );
+});
+
+const defaultResponseHeaders = { "Content-Type": "application/json" };
+
+afterEach(() => fetchMock.restore());
 
 describe('github API', () => {
-  const mockAPI = (api, responses) => {
-    api.request = (path, options = {}) => {
-      const normalizedPath = path.indexOf('?') !== -1 ? path.substr(0, path.indexOf('?')) : path;
-      const response = responses[normalizedPath];
-      return typeof response === 'function'
-        ? Promise.resolve(response(options))
-        : Promise.reject(new Error(`No response for path '${normalizedPath}'`))
-    };
-  }
+  it('should correctly detect a contributor', () => {
+    const api = new API({ branch: 'test-branch', repo: 'test-user/test-repo' });
+    mockRequest(api.api_root)('/user/repos', {
+      headers: defaultResponseHeaders,
+      body: [{
+        id: 1,
+        full_name: 'test-user/test-repo',
+        permissions: {
+          admin: false,
+          push: true,
+          pull: true,
+        },
+      }],
+    });
+    return expect(api.isCollaborator({})).resolves.toBe(true);
+  });
+  
+  it('should correctly detect a non-contributor', () => {
```

Comment:
Same as above. (Would it be more accurate to say "non-collaborator" instead of "non-contributor", since that is what we use in the code (and what GitHub uses)?)